core/src/main/java/jenkins/security/s2m/AdminWhitelistRule.java

```diff
@@ -160,19 +160,19 @@ public boolean checkFileAccess(String op, File f) {
 
     @RequirePOST
     public HttpResponse doSubmit(StaplerRequest req) throws IOException {
-        String whitelist = Util.fixNull(req.getParameter("whitelist"));
-        if (!whitelist.endsWith("\n"))
-            whitelist+="\n";
+        StringBuilder whitelist = new StringBuilder(Util.fixNull(req.getParameter("whitelist")));
+        if (!whitelist.toString().endsWith("\n"))
```

Comment:
You should be able to avoid constructing the string temporarily here by using `whitelist.lastIndexOf("\n") == (whitelist.length() - 1)` or something similar.